### PR TITLE
Iterate maps - for loop

### LINKED ISSUE #110: Add for loop to iterate maps
like `for: k, v in map.items()`
or something like that

### PR DESCRIPTION
**Pull request by Bonczidai**
fixes #110 

Signed-off-by: Bonczidai Levente <levente.bonczidai@hp.com>

### CODE REVIEW

**Comment by Bonczidai** [comment]
@orius123 please review

**Comment by orius123** [comment]
Sorry for responding so late by i think we stick with python
`for k, v in map.items()`
and not:
`for (k v) in map.items()`
WDYT?

**Comment by Bonczidai** [comment]
Fits for me, I will update the code

**Comment by Bonczidai** [comment]
updated to use `for k, v in map.items()`

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java [hunk 1]
```diff
@@ -21,7 +21,10 @@
 @Component
 public class ForTransformer implements Transformer<String, ForLoopStatement>{
 
+    // case: value in variable_name
     private final static String FOR_REGEX = "^(\\s+)?(\\w+)\\s+(in)\\s+(\\w+)(\\s+)?$";
+    // case: (key value)
+    private final static String KEY_VALUE_PAIR_REGEX = "^(\\s+)?(\\w+)(, )(\\w+)(\\s+)?$";
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java]: I would add optional spaces before and after the comma `^(\\s+)?(\\w+)(\\s+)?(,)(\\s+)?(\\w+)(\\s+)?$`
to match also `k  ,v` or `k,v` ..

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java [hunk 2]
```diff
@@ -60,10 +62,26 @@ public void incrementForLoop(String varName, Context flowContext, ForLoopConditi
...
 
-    private LoopCondition createForLoopCondition(String collectionExpression, Context flowContext, String nodeName) {
+    public void incrementMapForLoop(String keyName, String valueName, Context flowContext, ForLoopCondition forLoopCondition) {
+        Validate.notNull(keyName, "loop key name cannot be null");
+        Validate.notNull(keyName, "loop value name cannot be null");
+        Validate.notNull(flowContext, "flow context cannot be null");
+        Validate.notNull(forLoopCondition, "for condition cannot be null");
+
+        PyTuple keyValueTuple = (PyTuple) forLoopCondition.next();
+        Serializable keyFromIteration = (Serializable) keyValueTuple.get(0);
```
Comment (orius123) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: I would add trim

File: score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java [hunk 3]
```diff
@@ -30,22 +33,45 @@ public ForLoopStatement transform(String rawData) {
...
+            String beforeInKeyword = StringUtils.substringBefore(rawData, FOR_IN_KEYWORD);
+            collectionExpression = StringUtils.substringAfter(rawData, FOR_IN_KEYWORD).trim();
+
+            Pattern regexKeyValueFor = Pattern.compile(KEY_VALUE_PAIR_REGEX);
+            Matcher matcherKeyValueFor = regexKeyValueFor.matcher(beforeInKeyword);
+
+            if (matcherKeyValueFor.find()) {
+                // case: (key value)
+                String keyName = matcherKeyValueFor.group(2);
+                String valueName = matcherKeyValueFor.group(4);
```
Comment (orius123) [inline score-lang-compiler/src/main/java/org/openscore/lang/compiler/modeller/transformers/ForTransformer.java]: I think that if we have here the key and the value separated, it is a shame that we concat them with and a comma and then split them again at runtime.
I would prefer having to different constructors to the `ForLoopStatement` or having some subclass with the additional map data. 

and by that we can also not enforce the user to pass the `Type` eachtime
WDYT?

**Comment by orius123** [comment]
Added my comments.
In general, amazing job! really neat and tested.
and a great feature.
:100:

**Comment by meirwah** [comment]
To merge this?

File: score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java (hunk 2 above)
Comment (Bonczidai) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: What do you mean by adding trim? Where?

Comment (orius123) [inline score-lang-runtime/src/main/java/org/openscore/lang/runtime/bindings/LoopsBinding.java]: my bad. not here. in the transformer.

**Comment by Bonczidai** [comment]
@orius123 updated according to your comments..

I made ForLoopStatement abstract and created two subclasses for the two cases: list(normal) and maps

File: score-lang-compiler/src/test/java/org/openscore/lang/compiler/scorecompiler/ExecutionStepFactoryTest.java [hunk 4]
```diff
@@ -51,14 +51,14 @@ public void testCreateStartStepPutInputsUnderTheRightKey() throws Exception {
...
     public void testCreateStartStepPutForUnderTheRightKey() throws Exception {
-        ForLoopStatement statement = new ForLoopStatement("1", "2");
-        HashMap<String, Serializable> preTaskData = new HashMap<>();
-        preTaskData.put(SlangTextualKeys.FOR_KEY, statement);
-        ExecutionStep startStep = factory.createBeginTaskStep(1L, new ArrayList<Input>(), preTaskData, "", "");
-        ForLoopStatement actualStatement = (ForLoopStatement) startStep.getActionData()
-                                 .get(ScoreLangConstants.LOOP_KEY);
-        Assert.assertNotNull("for key is null", actualStatement);
-        Assert.assertSame("inputs are not set under their key", statement, actualStatement);
+//        ForLoopStatement statement = new ForLoopStatement("1", "2");
```
Comment (orius123) [inline score-lang-compiler/src/test/java/org/openscore/lang/compiler/scorecompiler/ExecutionStepFactoryTest.java]: ??

Comment (Bonczidai) [inline score-lang-compiler/src/test/java/org/openscore/lang/compiler/scorecompiler/ExecutionStepFactoryTest.java]: waoh.. my bad.. updating

File: score-lang-compiler/src/test/java/org/openscore/lang/compiler/CompileLoopsFlowTest.java [hunk 5]
```diff
@@ -9,6 +9,7 @@
  *******************************************************************************/
 package org.openscore.lang.compiler;
 
+import junit.framework.Assert;
```
Comment (orius123) [inline score-lang-compiler/src/test/java/org/openscore/lang/compiler/CompileLoopsFlowTest.java]: I always prefer org.junit

**Comment by orius123** [comment]
great job!! 
just add the test back and I will merge.

**Comment by Bonczidai** [comment]
thx..

done

**Pull request merged**